File: tests/rosetta/transpiler/FS/bitmap-b-zier-curves-cubic.fs
// Generated 2025-07-26 04:38 +0700

exception Break
exception Continue

exception Return

type Pixel = {
    r: int
    g: int
    b: int
}
let b3Seg: int = 30
let rec pixelFromRgb (rgb: int) =
    let mutable __ret : Pixel = Unchecked.defaultof<Pixel>
    let mutable rgb = rgb
    try
        let r: int = int ((((rgb / 65536) % 256 + 256) % 256))
        let g: int = int ((((rgb / 256) % 256 + 256) % 256))
        let b: int = int (((rgb % 256 + 256) % 256))
        __ret <- { r = r; g = g; b = b }
        raise Return
        __ret
    with
        | Return -> __ret
and newBitmap (cols: int) (rows: int) =
    let mutable __ret : Map<string, obj> = Unchecked.defaultof<Map<string, obj>>
    let mutable cols = cols
    let mutable rows = rows
    try
        let mutable d: Pixel array array = [||]
        let mutable y: int = 0
        while y < rows do
            let mutable row: Pixel array = [||]
            let mutable x: int = 0
            while x < cols do
                row <- Array.append row [|{ r = 0; g = 0; b = 0 }|]
                x <- x + 1
            d <- Array.append d [|row|]
            y <- y + 1
        __ret <- Map.ofList [("cols", box cols); ("rows", box rows); ("data", box d)]
        raise Return
        __ret
    with
        | Return -> __ret
and setPx (b: Map<string, obj>) (x: int) (y: int) (p: Pixel) =
    let mutable __ret : unit = Unchecked.defaultof<unit>
    let mutable b = b
    let mutable x = x
    let mutable y = y
    let mutable p = p
    try
        let cols: int = int (b.["cols"])
        let rows: int = int (b.["rows"])
        if (((x >= 0) && (x < cols)) && (y >= 0)) && (y < rows) then
            b <- Map.add "data" (Map.add y (Map.add x p ((b.["data"]).[y])) (b.["data"])) b
        __ret
    with
        | Return -> __ret
and fill (b: Map<string, obj>) (p: Pixel) =
    let mutable __ret : unit = Unchecked.defaultof<unit>
    let mutable b = b
    let mutable p = p
    try
        let cols: int = int (b.["cols"])
        let rows: int = int (b.["rows"])
        let mutable y: int = 0
        while y < rows do
            let mutable x: int = 0
            while x < cols do
                b <- Map.add "data" (Map.add y (Map.add x p ((b.["data"]).[y])) (b.["data"])) b
                x <- x + 1
            y <- y + 1
        __ret
    with
        | Return -> __ret
and fillRgb (b: Map<string, obj>) (rgb: int) =
    let mutable __ret : unit = Unchecked.defaultof<unit>
    let mutable b = b
    let mutable rgb = rgb
    try
        fill b (unbox<Pixel> (pixelFromRgb rgb))
        __ret
    with
        | Return -> __ret
and line (b: Map<string, obj>) (x0: int) (y0: int) (x1: int) (y1: int) (p: Pixel) =
    let mutable __ret : unit = Unchecked.defaultof<unit>
    let mutable b = b
    let mutable x0 = x0
    let mutable y0 = y0
    let mutable x1 = x1
    let mutable y1 = y1
    let mutable p = p
    try
        let mutable dx: int = x1 - x0
        if dx < 0 then
            dx <- -dx
        let mutable dy: int = y1 - y0
        if dy < 0 then
            dy <- -dy
        let mutable sx: int = -1
        if x0 < x1 then
            sx <- 1
        let mutable sy: int = -1
        if y0 < y1 then
            sy <- 1
        let mutable err: int = dx - dy
        try
            while true do
                setPx b x0 y0 p
                if (x0 = x1) && (y0 = y1) then
                    raise Break
                let e2: int = 2 * err
                if e2 > (0 - dy) then
                    err <- err - dy
                    x0 <- x0 + sx
                if e2 < dx then
                    err <- err + dx
                    y0 <- y0 + sy
        with
        | Break -> ()
        | Continue -> ()
        __ret
    with
        | Return -> __ret
and bezier3 (b: Map<string, obj>) (x1: int) (y1: int) (x2: int) (y2: int) (x3: int) (y3: int) (x4: int) (y4: int) (p: Pixel) =
    let mutable __ret : unit = Unchecked.defaultof<unit>
    let mutable b = b
    let mutable x1 = x1
    let mutable y1 = y1
    let mutable x2 = x2
    let mutable y2 = y2
    let mutable x3 = x3
    let mutable y3 = y3
    let mutable x4 = x4
    let mutable y4 = y4
    let mutable p = p
    try
        let mutable px: int array = [||]
        let mutable py: int array = [||]
        let mutable i: int = 0
        while i <= b3Seg do
            px <- Array.append px [|0|]
            py <- Array.append py [|0|]
            i <- i + 1
        let fx1: float = float x1
        let fy1: float = float y1
        let fx2: float = float x2
        let fy2: float = float y2
        let fx3: float = float x3
        let fy3: float = float y3
        let fx4: float = float x4
        let fy4: float = float y4
        i <- 0
        while i <= b3Seg do
            let d: float = (float i) / (float b3Seg)
            let mutable a: float = 1.0 - d
            let mutable bcoef: float = a * a
            let mutable ccoef: float = d * d
            let mutable a2: float = a * bcoef
            let mutable b2: float = (3.0 * bcoef) * d
            let mutable c2: float = (3.0 * a) * ccoef
            let mutable d2: float = ccoef * d
            px.[i] <- int ((((a2 * fx1) + (b2 * fx2)) + (c2 * fx3)) + (d2 * fx4))
            py.[i] <- int ((((a2 * fy1) + (b2 * fy2)) + (c2 * fy3)) + (d2 * fy4))
            i <- i + 1
        let mutable x0: int = px.[0]
        let mutable y0: int = py.[0]
        i <- 1
        while i <= b3Seg do
            let x: int = px.[i]
            let y: int = py.[i]
            line b x0 y0 x y p
            x0 <- x
            y0 <- y
            i <- i + 1
        __ret
    with
        | Return -> __ret
let mutable b: Map<string, obj> = newBitmap 400 300
fillRgb b 16773055
bezier3 b 20 200 700 50 (-300) 50 380 150 (unbox<Pixel> (pixelFromRgb 4165615))

File: tests/rosetta/transpiler/Kotlin/bitmap-b-zier-curves-cubic.kt
import java.math.BigInteger

var _nowSeed = 0L
var _nowSeeded = false
fun _now(): Long {
    if (!_nowSeeded) {
        System.getenv("MOCHI_NOW_SEED")?.toLongOrNull()?.let {
            _nowSeed = it
            _nowSeeded = true
        }
    }
    return if (_nowSeeded) {
        _nowSeed = (_nowSeed * 1664525 + 1013904223) % 2147483647
        kotlin.math.abs(_nowSeed)
    } else {
        kotlin.math.abs(System.nanoTime())
    }
}

fun toJson(v: Any?): String = when (v) {
    null -> "null"
    is String -> "\"" + v.replace("\"", "\\\"") + "\""
    is Boolean, is Number -> v.toString()
    is Map<*, *> -> v.entries.joinToString(prefix = "{", postfix = "}") { toJson(it.key.toString()) + ":" + toJson(it.value) }
    is Iterable<*> -> v.joinToString(prefix = "[", postfix = "]") { toJson(it) }
    else -> toJson(v.toString())
}

data class Pixel(var r: Int = 0, var g: Int = 0, var b: Int = 0)
var b3Seg: Int = 30
var b: MutableMap<String, Any?> = newBitmap(400, 300)
fun pixelFromRgb(rgb: Int): Pixel {
    var r: Int = (Math.floorMod((rgb / 65536), 256)).toInt()
    var g: Int = (Math.floorMod((rgb / 256), 256)).toInt()
    var b: Int = (Math.floorMod(rgb, 256)).toInt()
    return Pixel(r = r, g = g, b = b)
}

fun newBitmap(cols: Int, rows: Int): MutableMap<String, Any?> {
    var d: MutableList<MutableList<Pixel>> = mutableListOf<MutableList<Pixel>>()
    var y: Int = 0
    while (y < rows) {
        var row: MutableList<Pixel> = mutableListOf<Pixel>()
        var x: Int = 0
        while (x < cols) {
            row = run { val _tmp = row.toMutableList(); _tmp.add(Pixel(r = 0, g = 0, b = 0)); _tmp } as MutableList<Pixel>
            x = x + 1
        }
        d = run { val _tmp = d.toMutableList(); _tmp.add(row); _tmp } as MutableList<MutableList<Pixel>>
        y = y + 1
    }
    return mutableMapOf<String, Any?>("cols" to (cols), "rows" to (rows), "data" to (d))
}

fun setPx(b: MutableMap<String, Any?>, x: Int, y: Int, p: Pixel): Unit {
    var cols: Int = (b)["cols"] as Int
    var rows: Int = (b)["rows"] as Int
    if ((((((x >= 0) && (x < cols) as Boolean)) && (y >= 0) as Boolean)) && (y < rows)) {
        (((b)["data"] as Any? as MutableList<Any?>)[y]!! as MutableList<Any?>)[x] = p
    }
}

fun fill(b: MutableMap<String, Any?>, p: Pixel): Unit {
    var cols: Int = (b)["cols"] as Int
    var rows: Int = (b)["rows"] as Int
    var y: Int = 0
    while (y < rows) {
        var x: Int = 0
        while (x < cols) {
            (((b)["data"] as Any? as MutableList<Any?>)[y]!! as MutableList<Any?>)[x] = p
            x = x + 1
        }
        y = y + 1
    }
}

fun fillRgb(b: MutableMap<String, Any?>, rgb: Int): Unit {
    fill(b, pixelFromRgb(rgb))
}

fun line(b: MutableMap<String, Any?>, x0: Int, y0: Int, x1: Int, y1: Int, p: Pixel): Unit {
    var y0: Int = y0
    var x0: Int = x0
    var dx: BigInteger = (x1 - x0).toBigInteger()
    if (dx.compareTo((0).toBigInteger()) < 0) {
        dx = (0).toBigInteger().subtract((dx))
    }
    var dy: BigInteger = (y1 - y0).toBigInteger()
    if (dy.compareTo((0).toBigInteger()) < 0) {
        dy = (0).toBigInteger().subtract((dy))
    }
    var sx: Int = 0 - 1
    if (x0 < x1) {
        sx = 1
    }
    var sy: Int = 0 - 1
    if (y0 < y1) {
        sy = 1
    }
    var err: BigInteger = dx.subtract((dy))
    while (true) {
        setPx(b, x0, y0, p)
        if ((x0 == x1) && (y0 == y1)) {
            break
        }
        var e2: BigInteger = (2).toBigInteger().multiply((err))
        if (e2.compareTo(((0).toBigInteger().subtract((dy)))) > 0) {
            err = err.subtract((dy))
            x0 = x0 + sx
        }
        if (e2.compareTo((dx)) < 0) {
            err = err.add((dx))
            y0 = y0 + sy
        }
    }
}

fun bezier3(b: MutableMap<String, Any?>, x1: Int, y1: Int, x2: Int, y2: Int, x3: Int, y3: Int, x4: Int, y4: Int, p: Pixel): Unit {
    var px: MutableList<Int> = mutableListOf<Int>()
    var py: MutableList<Int> = mutableListOf<Int>()
    var i: Int = 0
    while (i <= b3Seg) {
        px = run { val _tmp = px.toMutableList(); _tmp.add(0); _tmp } as MutableList<Int>
        py = run { val _tmp = py.toMutableList(); _tmp.add(0); _tmp } as MutableList<Int>
        i = i + 1
    }
    var fx1: Double = x1.toDouble()
    var fy1: Double = y1.toDouble()
    var fx2: Double = x2.toDouble()
    var fy2: Double = y2.toDouble()
    var fx3: Double = x3.toDouble()
    var fy3: Double = y3.toDouble()
    var fx4: Double = x4.toDouble()
    var fy4: Double = y4.toDouble()
    i = 0
    while (i <= b3Seg) {
        var d: Double = (i.toDouble()) / (b3Seg.toDouble())
        var a: Double = 1.0 - d
        var bcoef: Double = a * a
        var ccoef: Double = d * d
        var a2: Double = a * bcoef
        var b2: Double = (3.0 * bcoef) * d
        var c2: Double = (3.0 * a) * ccoef
        var d2: Double = ccoef * d
        px[i] = ((((a2 * fx1) + (b2 * fx2)) + (c2 * fx3)) + (d2 * fx4)).toInt()
        py[i] = ((((a2 * fy1) + (b2 * fy2)) + (c2 * fy3)) + (d2 * fy4)).toInt()
        i = i + 1
    }
    var x0: Int = px[0]!!
    var y0: Int = py[0]!!
    i = 1
    while (i <= b3Seg) {
        var x: Int = px[i]!!
        var y: Int = py[i]!!
        line(b, x0, y0, x, y, p)
        x0 = x
        y0 = y
        i = i + 1
    }
}

fun main() {
    run {
        System.gc()
        val _startMem = Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory()
        val _start = _now()
        fillRgb(b, 16773055)
        bezier3(b, 20, 200, 700, 50, 0 - 300, 50, 380, 150, pixelFromRgb(4165615))
        System.gc()
        val _end = _now()
        val _endMem = Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory()
        val _durationUs = (_end - _start) / 1000
        val _memDiff = kotlin.math.abs(_endMem - _startMem)
        val _res = mapOf("duration_us" to _durationUs, "memory_bytes" to _memDiff, "name" to "main")
        println(toJson(_res))
    }
}
